### DISCUSSION TITLE
Where is HAVE_SQLITE3_COLUMN_DATABASE_NAME defined?

### DISCUSSION THREAD

**Discussion by fractaledmind** [sparklemotion/sqlite3-ruby#464] [Q&A] [answered]
I am trying to understand how the gem determines whether or not to include the `database_name` method in the definition of the `Statement` class. 

The code references a `HAVE_SQLITE3_COLUMN_DATABASE_NAME` constant in two places:
https://github.com/sparklemotion/sqlite3-ruby/blob/536152860daf4e4acbf0608d0840941f1a878722/ext/sqlite3/statement.c#L414
https://github.com/sparklemotion/sqlite3-ruby/blob/536152860daf4e4acbf0608d0840941f1a878722/ext/sqlite3/statement.c#L452

And the tests say that this method is defined or not depending on how SQLite was compiled:
https://github.com/sparklemotion/sqlite3-ruby/blob/536152860daf4e4acbf0608d0840941f1a878722/test/test_statement.rb#L48

But, [a search](https://github.com/search?q=repo%3Asqlite%2Fsqlite+HAVE_SQLITE3_COLUMN_DATABASE_NAME&type=code) of the SQLite source doesn't return any use of this constant. And [the docs](https://www.sqlite.org/compile.html#enable_column_metadata) say that the compilation flag used is `SQLITE_ENABLE_COLUMN_METADATA`. 

So, how and where is the `HAVE_SQLITE3_COLUMN_DATABASE_NAME` constant set and how does that relate to the `SQLITE_ENABLE_COLUMN_METADATA` compilation flag?

**Comment by flavorjones** [comment, marked as answer]
Happy to answer questions! Let's do it in Discussions, though, and not Issues.

In the extconf (`ext/sqlite3/extconf.rb`), which is where the C extension is configured, you'll see this line:

```ruby
        have_func('sqlite3_column_database_name')
```

The method `have_func` comes from Ruby's `mkmf.rb` or `MakeMakefile` module, and is documented here:

https://docs.ruby-lang.org/en/master/MakeMakefile.html#method-i-have_func

Relevant:

> Returns whether or not the function `func` can be found in the common header files, or within any `headers` that you provide. If found, a macro is passed as a preprocessor constant to the compiler using the function name, in uppercase, prepended with `HAVE_`.
>
> For example, if `have_func('foo')` returned `true`, then the `HAVE_FOO` preprocessor macro would be passed to the compiler.

Make sense?

**Reply by fractaledmind** [reply]
Makes perfect sense! And sorry for forgetting about Discussions again; still haven't formed the habit, but I'll get it switched on.

Seems this should be how I handle the `progress_handler` as well, as it can be present or not in the compiled SQLite source depending on flags.

Thanks for the clear answer!
